### ISSUE TITLE
ActiveStorage: Support AWS IAM Roles when using S3

### ISSUE THREAD

**Issue by tbmetcalfe** [rails/rails#48836]
### Description

Currently, it is mandatory to use AWS Access Keys when using ActiveStorage. However, it is considered best practice to use AWS IAM roles where possible if operating within AWS. Currently, this does not appear possible when using ActiveStorage.

### Current behavior
If you do not set these fields, `access_key_id` and `secret_access_key`, when configuring to use S3, then even when you set the other methods of setting AWS credentials, the service will fail as it cannot locate the credentials. 

### Desired behavior
If the `access_key_id` and `secret_access_key` are set, then use these fields. If, however, these fields are not set, it will then try to locate the default AWS credentials (as defined in the AWS SDK) such as from environment variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`) or the IAM role assigned to the resource (e.g. if it's running on an EC2 instance)

This behavior is already implemented in the AWS SDK for all programming languages.

### System configuration
**Rails version**: `6.1.3`

**Ruby version**: `3.0.5`

**Comment by yahonda** [comment]
Feel free to open a pull request.

**Comment by Jay-Ang** [comment]
@tbmetcalfe I've written some basic tests, and everything appears to be functioning correctly on my end. If the `access_key_id` and `secret_access_key` do not exist or are empty in the configuration file, the AWS SDK will automatically fall back to using environment variables. Please let me know if there are any other details I should consider.

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "6.1.3"
  gem "aws-sdk-s3", "~> 1.48"
end

require 'active_support/core_ext/array/access'
require 'aws-sdk-s3'
require "active_storage"
require "logger"
require "minitest/autorun"

ActiveRecord::Base.logger = Logger.new(STDOUT)

class S3ServiceTest < Minitest::Test
  def setup
    @key  = "test.txt"
    @data = "Something else entirely!"
    @checksum = OpenSSL::Digest::MD5.base64digest(@data)
    @service_configuration = {
      :s3 => {
        :service => "S3",
        :access_key_id => "",
        :secret_access_key => "",
        :bucket => "rails",
        :region => "us-east-1",
      }
    }
  end

  def teardown
    ENV["AWS_ACCESS_KEY_ID"] = nil
    ENV["AWS_SECRET_ACCESS_KEY"] = nil
  end

  def test_without_credentials
    assert_raises Aws::Errors::MissingCredentialsError do
      upload_file
    end
  end

  # When the credentials are empty in the configuration, the AWS SDK will
  # fallback to the environment variables. Thus, we will get InvalidAccessKeyId error.
  def test_with_enviroment_variables_credentials
    ENV["AWS_ACCESS_KEY_ID"] = "123"
    ENV["AWS_SECRET_ACCESS_KEY"] = "123"

    assert_raises Aws::S3::Errors::InvalidAccessKeyId do
      upload_file
    end
  end

  private

  def upload_file
    service = ActiveStorage::Service.configure(:s3, @service_configuration)
    service.upload(@key, StringIO.new(@data), checksum: @checksum)
  end
end
```

**Comment by Jay0921** [comment]
@tbmetcalfe I've written some basic tests, and everything appears to be functioning correctly on my end. If the `access_key_id` and `secret_access_key` do not exist or are empty in the configuration file, the AWS SDK will automatically fall back to using environment variables. Please let me know if there are any other details I should consider.

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "6.1.3"
  gem "aws-sdk-s3", "~> 1.48"
end

require 'active_support/core_ext/array/access'
require 'aws-sdk-s3'
require "active_storage"
require "logger"
require "minitest/autorun"

ActiveRecord::Base.logger = Logger.new(STDOUT)

class S3ServiceTest < Minitest::Test
  def setup
    @key  = "test.txt"
    @data = "Something else entirely!"
    @checksum = OpenSSL::Digest::MD5.base64digest(@data)
    @service_configuration = {
      :s3 => {
        :service => "S3",
        :access_key_id => "",
        :secret_access_key => "",
        :bucket => "rails",
        :region => "us-east-1",
      }
    }
  end

  def teardown
    ENV["AWS_ACCESS_KEY_ID"] = nil
    ENV["AWS_SECRET_ACCESS_KEY"] = nil
  end

  def test_without_credentials
    assert_raises Aws::Errors::MissingCredentialsError do
      upload_file
    end
  end

  # When the credentials are empty in the configuration, the AWS SDK will
  # fallback to the environment variables. Thus, we will get InvalidAccessKeyId error.
  def test_with_enviroment_variables_credentials
    ENV["AWS_ACCESS_KEY_ID"] = "123"
    ENV["AWS_SECRET_ACCESS_KEY"] = "123"

    assert_raises Aws::S3::Errors::InvalidAccessKeyId do
      upload_file
    end
  end

  private

  def upload_file
    service = ActiveStorage::Service.configure(:s3, @service_configuration)
    service.upload(@key, StringIO.new(@data), checksum: @checksum)
  end
end
```

**Comment by morgoth** [comment]
@tbmetcalfe I'm also using the active storage with S3 configured via IAM role, thus without secret/key config and it works for me.
Can this be closed?

**Comment by tbmetcalfe** [comment]
> morgoth

@morgoth thank you for confirming this, this is really useful to know! I've checked it out in my own project and still can't get this to work. Can you tell me what version of rails and ruby you're running? Perhaps some docs on this issue could be added? I couldn't find any information regarding IAM roles.

**Comment by morgoth** [comment]
I'm using Rails main branch, so the upcoming 7.1.
Ruby 3.2.2

**Comment by tbmetcalfe** [comment]
For anyone reading this, I discovered this was not a rails issue but rather an NGINX misconfiguration. 

NGINX was returning 502s because the signed URL generated by active storage using AWS IAM roles was much longer than one generated by AWS Access Keys which broke the default NGINX `proxy_buffers` and `proxy_buffer_size` constraint. 

I'm running this on the NGINX ingress controller so I added the following annotations to fix this:

```yaml
nginx.ingress.kubernetes.io/proxy-buffer-size: 128k
nginx.ingress.kubernetes.io/proxy-buffers: "4 128k"
```

The NGINX config would be:
```
http {
  proxy_buffer_size   128k;
  proxy_buffers   4 256k;
}
```

Cheers @morgoth for letting me know that this was possible!
